File: web/src/engine/stats.ts
/**
 * 统计原语（对应 Python `src/koxpilot/stats.py`）。
 *
 * 双实现一致性的关键在这一层：Python 与 TS 必须做**逐位相同**的浮点运算，
 * 否则 authenticity_score 在 0.5 的判定边界上会出现零星错判，
 * 而这种差异排查起来非常昂贵。因此这里刻意做了三件事：
 *  1. `pyRound` 复现 CPython `round(x, n)` 的"精确值 + 银行家进位"语义
 *     （JS 原生 `toFixed` 在恰好压线时是 half-up，与 Python 不同）；
 *  2. 求和一律保持与 Python 相同的**累加顺序**（builtin sum = 从左到右）；
 *  3. 需要 `math.fsum`（精确求和）的地方用 Neumaier 补偿求和近似，并在注释里标注。
 */

const EPS = 1e-12;

/** 除零保护除法（对应 `safe_div`）。 */
export function safeDiv(numerator: number, denominator: number, dflt = 0.0): number {
  if (denominator === null || denominator === undefined || Math.abs(denominator) < EPS) return dflt;
  return numerator / denominator;
}

/**
 * CPython `round(x, ndigits)` 的等价实现。
 *
 * 非压线情况下 `Number(x.toFixed(n))` 与 Python 结果同为"最接近该十进制值的 double"，二者一致；
 * 唯一分歧是**精确压线**（如 round(0.03125, 4)：Python→0.0312 取偶，toFixed→0.0313 进位）。
 * 双精度是二进制小数，十进制展开有限，因此用 20 位定点串即可判定是否精确压线。
 */
export function pyRound(x: number, n = 0): number {
  if (!Number.isFinite(x)) return x;
  const neg = x < 0;
  const a = Math.abs(x);
  if (a >= 1e15) return x; // 该量级下 double 已无小数精度，无需处理
  const s = a.toFixed(20);
  const dot = s.indexOf('.');
  const intPart = s.slice(0, dot);
  const dec = s.slice(dot + 1);
  const tail = dec.slice(n);
  let out: number;
  const isTie = tail.charAt(0) === '5' && /^0*$/.test(tail.slice(1));
  if (isTie) {
    const keep = dec.slice(0, n);
    const truncated = Number(n > 0 ? `${intPart}.${keep}` : intPart);
    const lastDigit = Number(n > 0 ? keep.charAt(n - 1) || '0' : intPart.charAt(intPart.length - 1));
    if (lastDigit % 2 === 0) {
      out = truncated; // 已是偶数 -> 舍
    } else {
      out = Number((truncated + Math.pow(10, -n)).toFixed(n)); // 进位到偶数
    }
  } else {
    out = Number(a.toFixed(n));
  }
  return neg ? -out : out;
}

/** Neumaier 补偿求和，近似 Python `math.fsum`（只用于 mean 的退化分支）。 */
export function fsum(values: readonly number[]): number {
  let s = 0.0;
  let c = 0.0;
  for (const v of values) {
    const t = s + v;
    if (Math.abs(s) >= Math.abs(v)) c += s - t + v;
    else c += v - t + s;
    s = t;
  }
  return s + c;
}

export function mean(values: readonly number[]): number {
  if (values.length === 0) return 0.0;
  return fsum(values) / values.length;
}

/**
 * 线性插值分位数（等价 numpy 'linear' / R type-7），与 Python `quantile` 逐步对应。
 * 入参不要求有序，函数内部排序且不修改入参。
 */
export function quantile(values: readonly number[], q: number): number {
  if (!(q >= 0.0 && q <= 1.0)) throw new Error(`q must be in [0, 1], got ${q}`);
  if (values.length === 0) return 0.0;
  const ordered = [...values].sort((a, b) => a - b);
  const n = ordered.length;
  if (n === 1) return ordered[0];
  const h = (n - 1) * q;
  const lo = Math.floor(h);
  const hi = Math.ceil(h);
  if (lo === hi) return ordered[lo];
  const frac = h - lo;
  return ordered[lo] * (1.0 - frac) + ordered[hi] * frac;
}

export function median(values: readonly number[]): number {
  return quantile(values, 0.5);
}

/** 中位数绝对偏差（G1.5 突刺检测的鲁棒尺度）。 */
export function mad(values: readonly number[]): number {
  if (values.length === 0) return 0.0;
  const med = median(values);
  return median(values.map((v) => Math.abs(v - med)));
}

/**
 * 鲁棒 z-score：`z = (x - median) / (1.4826 · MAD)`。
 * 1.4826 = 1/Φ⁻¹(0.75)，作用是让 MAD 在正态下与 σ 同尺度，
 * 这样 SPEC 的 "z > 2.5" 与常规 σ 语义可比。
 */
export function robustZscores(values: readonly number[]): number[] {
  if (values.length === 0) return [];
  const med = median(values);
  let scale = 1.4826 * mad(values);
  if (scale < EPS) scale = mean(values.map((v) => Math.abs(v - med)));
  if (scale < EPS) return values.map(() => 0.0);
  return values.map((v) => (v - med) / scale);
}

/** 集合 Jaccard；两个空集定义为 1.0（都没声明品类不算冲突）。 */
export function jaccard(a: Iterable<string>, b: Iterable<string>): number {
  const sa = new Set(a);
  const sb = new Set(b);
  if (sa.size === 0 && sb.size === 0) return 1.0;
  const union = new Set([...sa, ...sb]);
  if (union.size === 0) return 1.0;
  let inter = 0;
  for (const v of sa) if (sb.has(v)) inter += 1;
  return inter / union.size;
}

/** ROC-AUC（Mann–Whitney U 秩和 + 并列秩平均修正），供评测页在浏览器里重算。 */
export function rocAuc(scores: readonly number[], labels: readonly number[]): number {
  if (scores.length !== labels.length) throw new Error('scores/labels length mismatch');
  const n = scores.length;
  const nPos = labels.reduce((acc, y) => acc + (y === 1 ? 1 : 0), 0);
  const nNeg = n - nPos;
  if (nPos === 0 || nNeg === 0) return 0.5;
  const order = Array.from({ length: n }, (_, i) => i).sort((i, j) => scores[i] - scores[j]);
  const ranks = new Array<number>(n).fill(0);
  let i = 0;
  while (i < n) {
    let j = i;
    while (j + 1 < n && scores[order[j + 1]] === scores[order[i]]) j += 1;
    const avgRank = (i + j) / 2.0 + 1.0;
    for (let k = i; k <= j; k += 1) ranks[order[k]] = avgRank;
    i = j + 1;
  }
  const rPos = fsum(ranks.filter((_, idx) => labels[idx] === 1));
  return (rPos - (nPos * (nPos + 1)) / 2.0) / (nPos * nNeg);
}

export interface PRF1 {
  tp: number;
  fp: number;
  fn: number;
  precision: number;
  recall: number;
  f1: number;
}

/** 由混淆计数算 P/R/F1，分母为 0 的格子按 0 报（不隐藏）。 */
export function prf1(tp: number, fp: number, fn: number): PRF1 {
  const precision = safeDiv(tp, tp + fp);
  const recall = safeDiv(tp, tp + fn);
  const f1 = safeDiv(2 * precision * recall, precision + recall);
  return { tp, fp, fn, precision: pyRound(precision, 4), recall: pyRound(recall, 4), f1: pyRound(f1, 4) };
}

File: src/koxpilot/stats.py
"""KOXPilot 统计原语（全部手写，不依赖 sklearn/pandas）。

设计意图
--------
本模块是整个项目"阈值可解释"的地基。项目纪律要求：
1. 任何门禁阈值都必须能追溯到**数据分位数**或**显式配置**，不允许魔法数字；
2. 评测指标（AUC / P / R / F1）必须能用几句话讲清算法，
   因此不引入 sklearn，而是自己实现（AUC 用 Mann–Whitney U 的秩和公式，含并列秩修正）。

所有函数都是纯函数：输入序列 -> 输出数值，不读全局状态、不写文件。
"""

from __future__ import annotations

import math
from collections.abc import Iterable, Sequence
from typing import Final

__all__ = [
    "jaccard",
    "mad",
    "mean",
    "median",
    "percentile_rank",
    "prf1",
    "quantile",
    "quantiles",
    "robust_zscores",
    "roc_auc",
    "safe_div",
]

_EPS: Final[float] = 1e-12


def safe_div(numerator: float, denominator: float, default: float = 0.0) -> float:
    """除零保护除法。评测里到处需要它，集中一处实现避免各处写 if。"""
    if denominator is None or abs(denominator) < _EPS:
        return default
    return numerator / denominator


def mean(values: Sequence[float]) -> float:
    if not values:
        return 0.0
    return math.fsum(values) / len(values)


def quantile(values: Sequence[float], q: float) -> float:
    """线性插值分位数（等价于 numpy 默认的 'linear' / R 的 type-7）。

    手写原因：这是全项目所有 G1 阈值的唯一来源，必须能逐步复算核对。
    实现：把 q 映射到排序后数组的实数位置 h = (n-1)*q，
    再在 floor(h) 与 ceil(h) 两个样本之间线性插值。

    Args:
        values: 样本序列（不要求已排序，函数内部排序，不修改入参）。
        q: 分位点，取值 [0, 1]。
    Returns:
        分位数值。空序列返回 0.0（调用方需自行判断样本量是否足够）。
    """
    if not 0.0 <= q <= 1.0:
        raise ValueError(f"q must be in [0, 1], got {q}")
    if not values:
        return 0.0
    ordered = sorted(values)
    n = len(ordered)
    if n == 1:
        return float(ordered[0])
    h = (n - 1) * q
    lo = math.floor(h)
    hi = math.ceil(h)
    if lo == hi:
        return float(ordered[int(h)])
    frac = h - lo
    return float(ordered[lo] * (1.0 - frac) + ordered[hi] * frac)


def quantiles(values: Sequence[float], qs: Iterable[float]) -> dict[str, float]:
    """一次算多个分位数，key 形如 ``p05`` / ``p97`` / ``p02``。

    key 用两位（必要时三位）整数百分位表示，前端 TS 侧读同一份 JSON 时按同名 key 取值。
    """
    ordered = sorted(values)
    out: dict[str, float] = {}
    for q in qs:
        pct = q * 100.0
        label = f"p{int(round(pct)):02d}" if abs(pct - round(pct)) < 1e-9 else f"p{pct:g}"
        out[label] = quantile(ordered, q)
    return out


def median(values: Sequence[float]) -> float:
    return quantile(values, 0.5)


def mad(values: Sequence[float]) -> float:
    """Median Absolute Deviation（中位数绝对偏差）。

    用于 G1.5 粉丝突刺检测：单账号的月度增速序列很短（11 个点）且本身就含突刺，
    普通标准差会被突刺自己撑大导致检不出来（掩蔽效应），所以用 MAD 做鲁棒尺度。
    """
    if not values:
        return 0.0
    med = median(values)
    return median([abs(v - med) for v in values])


def robust_zscores(values: Sequence[float]) -> list[float]:
    """鲁棒 z-score：z_i = (x_i - median) / (1.4826 * MAD)。

    1.4826 = 1/Φ⁻¹(0.75)，作用是让 MAD 在正态分布下与标准差一致，
    这样 SPEC 里 "z > 2.5" 的门槛与常规 σ 语义可比。
    MAD 为 0 时（序列几乎恒定）退化为用均值绝对偏差，仍为 0 则返回全 0。
    """
    if not values:
        return []
    med = median(values)
    scale = 1.4826 * mad(values)
    if scale < _EPS:
        scale = mean([abs(v - med) for v in values])
    if scale < _EPS:
        return [0.0] * len(values)
    return [(v - med) / scale for v in values]


def percentile_rank(sorted_values: Sequence[float], x: float) -> float:
    """x 在已排序样本中的经验百分位（0~1），用二分查找。

    G1 的连续异常分（用于算 AUC）就是基于它构造的：
    把"实际值"映射成"在同平台同量级同行里排第几"，天然无量纲、跨组可比。
    """
    n = len(sorted_values)
    if n == 0:
        return 0.5
    lo, hi = 0, n
    while lo < hi:  # bisect_left
        mid = (lo + hi) // 2
        if sorted_values[mid] < x:
            lo = mid + 1
        else:
            hi = mid
    left = lo
    lo, hi = 0, n
    while lo < hi:  # bisect_right
        mid = (lo + hi) // 2
        if sorted_values[mid] <= x:
            lo = mid + 1
        else:
            hi = mid
    right = lo
    return (left + right) / (2.0 * n)


def jaccard(a: Iterable[str], b: Iterable[str]) -> float:
    """集合 Jaccard 相似度。两个空集定义为 1.0（都没声明品类不算冲突）。

    G2.1 / G2.2 的品类一致性判定基于它。定义边界写在这里以免各处口径不一。
    """
    sa, sb = set(a), set(b)
    if not sa and not sb:
        return 1.0
    union = sa | sb
    if not union:
        return 1.0
    return len(sa & sb) / len(union)


def roc_auc(scores: Sequence[float], labels: Sequence[int]) -> float:
    """ROC-AUC，用 Mann–Whitney U 秩和公式实现（含并列秩平均修正）。

        AUC = (R_pos - n_pos*(n_pos+1)/2) / (n_pos * n_neg)

    其中 R_pos 是正样本的秩和。含义：随机取一正一负，正样本分更高的概率
    （并列算 0.5）。这是可口述复算的版本：O(n log n)，不需要扫阈值画曲线。

    Args:
        scores: 连续预测分（越大越像正类）。
        labels: 0/1 真值。
    Returns:
        AUC。正类或负类为空时返回 0.5（无判别信息）。
    """
    if len(scores) != len(labels):
        raise ValueError("scores/labels length mismatch")
    n = len(scores)
    n_pos = sum(1 for y in labels if y == 1)
    n_neg = n - n_pos
    if n_pos == 0 or n_neg == 0:
        return 0.5
    order = sorted(range(n), key=lambda i: scores[i])
    ranks = [0.0] * n
    i = 0
    while i < n:
        j = i
        while j + 1 < n and scores[order[j + 1]] == scores[order[i]]:
            j += 1
        avg_rank = (i + j) / 2.0 + 1.0  # 秩从 1 开始，并列取平均
        for k in range(i, j + 1):
            ranks[order[k]] = avg_rank
        i = j + 1
    r_pos = math.fsum(ranks[i] for i in range(n) if labels[i] == 1)
    return (r_pos - n_pos * (n_pos + 1) / 2.0) / (n_pos * n_neg)


def prf1(tp: int, fp: int, fn: int) -> dict[str, float]:
    """由混淆计数算 precision / recall / f1。所有分母为 0 的格子按 0.0 报，不隐藏。"""
    precision = safe_div(tp, tp + fp)
    recall = safe_div(tp, tp + fn)
    f1 = safe_div(2 * precision * recall, precision + recall)
    return {
        "tp": tp,
        "fp": fp,
        "fn": fn,
        "precision": round(precision, 4),
        "recall": round(recall, 4),
        "f1": round(f1, 4),
    }
